server/athenian/api/controllers/features/github/pull_request.py

```diff
@@ -0,0 +1,212 @@
+    ns = 1_000_000_000
+    max_conf_max_ratio = 10
+    dtype_is_timedelta = isinstance(data[0], (pd.Timedelta, timedelta))
+    if dtype_is_timedelta:
+        # we have to convert the dtype because some ops required by scipy are missing
+        # thus the precision is 1 second; otherwise there are integer overflows
+        if isinstance(data[0], pd.Timedelta):
+            arr = np.asarray([d.to_timedelta64() // ns for d in data], dtype=np.int64)
+        else:
+            arr = np.asarray(data, dtype="timedelta64[ns]").astype(np.int64) // ns
+    else:
+        arr = np.asarray(data)
+    if may_have_negative_values:
+        # assume a normal distribution
+        m = np.mean(arr)
+        sem = scipy.stats.sem(arr)
+        if sem != sem:
+            # only one sample, we don't know the stddev so whatever value to indicate a failure
+            conf_min = type(m)(0)
+            conf_max = max_conf_max_ratio * m
+        else:
+            conf_min, conf_max = scipy.stats.t.interval(confidence, len(arr) - 1, loc=m, scale=sem)
+    else:
+        # assume a log-normal distribution
+        assert (arr >= 0).all()
+        arr = arr[arr != 0]
+        m = np.mean(arr)
+        if len(arr) == 1:
+            # only one sample, we don't know the stddev so whatever value to indicate a failure
+            conf_min = type(m)(0)
+            conf_max = max_conf_max_ratio * m
+        else:
+            # modified Cox method
+            # https://doi.org/10.1080/10691898.2005.11910638
+            logarr = np.log(arr)
+            logvar = np.var(logarr, ddof=1)
+            logm = np.mean(logarr)
+            d = np.sqrt(logvar / len(arr) + logvar**2 / (2 * (len(arr) - 1)))
+            conf_min, conf_max = scipy.stats.t.interval(
+                confidence, len(arr) - 1, loc=logm + logvar / 2, scale=d)
+            conf_min, conf_max = np.exp(conf_min), np.exp(conf_max)
+            if conf_max / m > max_conf_max_ratio:
+                conf_max = max_conf_max_ratio * m
+    if dtype_is_timedelta:
+        # convert the dtype back
+        m = pd.Timedelta(np.timedelta64(int(m * ns)))
+        conf_min = pd.Timedelta(np.timedelta64(int(conf_min * ns)))
+        conf_max = pd.Timedelta(np.timedelta64(int(conf_max * ns)))
+        if isinstance(data[0], timedelta):
+            m = m.to_pytimedelta()
+            conf_min = conf_min.to_pytimedelta()
+            conf_max = conf_max.to_pytimedelta()
+    else:
+        dt = type(data[0])
+        m = dt(m)
+        conf_min = dt(conf_min)
+        conf_max = dt(conf_max)
+    return m, conf_min, conf_max
+
+
+def median_confidence_interval(data: Sequence[T], confidence=0.95,
+                               ) -> Tuple[Optional[T], Optional[T], Optional[T]]:
+    """Calculate the median value and the confidence interval."""
+    if len(data) == 0:
+        return None, None, None
+    arr = np.asarray(data)
+    # https://onlinecourses.science.psu.edu/stat414/node/316
```

Comment:
I don't think we want to keep this url here ;)

Reply:
Heh, that doc explains the whole code block below.

Reply:
I'd keep it for future reference, hoping that the link will work.